server/events/project_command_runner.go

```diff
@@ -343,31 +341,11 @@ func (p *DefaultProjectCommandRunner) doApply(ctx models.ProjectCommandContext)
 		return "", "", DirNotExistErr{RepoRelDir: ctx.RepoRelDir}
 	}
 
-	for _, req := range ctx.ApplyRequirements {
-		switch req {
-		case raw.ApprovedApplyRequirement:
-			approved, err := p.PullApprovedChecker.PullIsApproved(ctx.Pull.BaseRepo, ctx.Pull) // nolint: vetshadow
-			if err != nil {
-				return "", "", errors.Wrap(err, "checking if pull request was approved")
-			}
-			if !approved {
-				return "", "Pull request must be approved by at least one person other than the author before running apply.", nil
-			}
-		// this should come before mergeability check since mergeability is a superset of this check.
-		case valid.PoliciesPassedApplyReq:
-			if ctx.ProjectPlanStatus == models.ErroredPolicyCheckStatus {
-				return "", "All policies must pass for project before running apply", nil
-			}
-		case raw.MergeableApplyRequirement:
-			if !ctx.PullMergeable {
-				return "", "Pull request must be mergeable before running apply.", nil
-			}
-		case raw.UnDivergedApplyRequirement:
-			if p.WorkingDir.HasDiverged(ctx.Log, repoDir) {
-				return "", "Default branch must be rebased onto pull request before running apply.", nil
-			}
-		}
+	failure, err = p.AggregateApplyRequirements.ValidateProject(repoDir, ctx)
+	if failure != "" || err != nil {
+		return "", failure, err
```

Comment:
It's best practice to just check the err object in go.  You should assume that the failure message will be populated as such by the underlying implementation.

so just:

```
if err != nil {
...

}


```

Reply:
In this case, I think we want to return from this function when either an apply requirement is not met which populates the failure message, or when an error occurs when trying to validate the apply requirement which populates the error. So, I think we should be checking for both the failure message and the error. We only want to continue with the execution in `doApply()` when all the apply requirements are met and there's no error (i.e when failure == "" and err == nil).

Reply:
The tests fail when I make that change, specifically the `TestDefaultProjectCommandRunner_ApplyNotApproved()` because we're expecting it to return from the function with a failure message `Pull request must be approved by at least one person other than the author before running apply.` but since the err is nil, we continue with the execution in `doApply()` which is throwing errors.

Reply:
yeah you're right I misread it.